src/shell/shell_misc.cpp

```diff
@@ -410,49 +409,38 @@ bool DOS_Shell::Execute(char * name,char * args) {
 		}
 		return true;
 	}
+
 	/* Check for a full name */
-	p_fullname = Which(name);
-	if (!p_fullname) return false;
+	const char *p_fullname = Which(name);
+	if (!p_fullname)
+		return false;
 	safe_strcpy(fullname, p_fullname);
-	const char* extension = strrchr(fullname,'.');
-	
-	/*always disallow files without extension from being executed. */
-	/*only internal commands can be run this way and they never get in this handler */
-	if(extension == 0)
-	{
-		//Check if the result will fit in the parameters. Else abort
-		if(strlen(fullname) >( DOS_PATHLENGTH - 1) ) return false;
-		char temp_name[DOS_PATHLENGTH+4],* temp_fullname;
-		//try to add .com, .exe and .bat extensions to filename
-
-		safe_strcpy(temp_name, fullname);
-		safe_strcat(temp_name, ".COM");
-		temp_fullname=Which(temp_name);
-		if (temp_fullname) { extension=".com";strcpy(fullname,temp_fullname); }
-
-		else 
-		{
-			safe_strcpy(temp_name, fullname);
-			safe_strcat(temp_name, ".EXE");
-			temp_fullname=Which(temp_name);
-		 	if (temp_fullname) { extension=".exe";strcpy(fullname,temp_fullname);}
 
-			else 
-			{
-				safe_strcpy(temp_name, fullname);
-				safe_strcat(temp_name, ".BAT");
-				temp_fullname=Which(temp_name);
-		 		if (temp_fullname) { extension=".bat";strcpy(fullname,temp_fullname);}
-
-				else  
-				{
-		 			return false;
-				}
-			
-			}	
+	// Always disallow files without extension from being executed.
+	// Only internal commands can be run this way and they never get in
+	// this handler.
+	const char *extension = strrchr(fullname, '.');
+	if (extension == nullptr) {
+		// Check if the result will fit in the parameters.
+		if (strlen(fullname) > (DOS_PATHLENGTH - 1))
+			return false;
+
+		// Try to add .COM, .EXE and .BAT extensions to the filename
+		char temp_name[DOS_PATHLENGTH + 4];
```

Comment:
If we can nail down the check above, then the only files left to work with are those that can accomodate the ".EXT", in which case we don't need the `+ 4` here.